client/trino-cli/src/main/java/io/trino/cli/MarkdownTablePrinter.java

```diff
@@ -0,0 +1,87 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.cli;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.client.Column;
+import net.steppschuh.markdowngenerator.table.Table;
+import net.steppschuh.markdowngenerator.table.TableRow;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.client.ClientStandardTypes.BIGINT;
+import static io.trino.client.ClientStandardTypes.DECIMAL;
+import static io.trino.client.ClientStandardTypes.DOUBLE;
+import static io.trino.client.ClientStandardTypes.INTEGER;
+import static io.trino.client.ClientStandardTypes.REAL;
+import static io.trino.client.ClientStandardTypes.SMALLINT;
+import static io.trino.client.ClientStandardTypes.TINYINT;
+import static java.lang.System.lineSeparator;
+import static java.util.Objects.requireNonNull;
+
+public class MarkdownTablePrinter
+        implements OutputPrinter
+{
+    private static final Set<String> NUMERIC_TYPES = ImmutableSet.of(TINYINT, SMALLINT, INTEGER, BIGINT, REAL, DOUBLE, DECIMAL);
+    private final List<String> fieldNames;
+    private final List<Integer> alignments;
+    private final Writer writer;
+
+    public MarkdownTablePrinter(List<Column> columns, Writer writer)
+    {
+        requireNonNull(columns, "columns is null");
+        this.fieldNames = columns.stream()
+                .map(Column::getName)
+                .collect(toImmutableList());
+        this.alignments = columns.stream()
+                .map(Column::getTypeSignature)
+                .map(signature -> NUMERIC_TYPES.contains(signature.getRawType()) ? Table.ALIGN_RIGHT : Table.ALIGN_LEFT)
+                .collect(toImmutableList());
+        this.writer = requireNonNull(writer, "writer is null");
+    }
+
+    @Override
+    public void printRows(List<List<?>> rows, boolean complete)
+            throws IOException
+    {
+        Table.Builder tableBuilder = new Table.Builder()
+                .withAlignments(alignments)
+                .addRow(new TableRow<>(fieldNames));
+        rows.forEach(row -> {
+            List<String> formattedRow = row.stream().map(MarkdownTablePrinter::formatValue).collect(Collectors.toList());
+            tableBuilder.addRow(new TableRow<>(formattedRow));
+        });
+        writer.append(tableBuilder.build().toString());
+        writer.append(lineSeparator());
```

Comment:
I believe that printRows can be called multiple times.

we should call writer.append(tableBuilder.build().toString()); in the `if (complete) {`

Reply:
It'll be called every 10k rows. I don't think we should be adding another buffer in the table printer. Printing that many rows is expensive, since widths needs to be calculated for all columns. This is out of the scope of this PR.

Reply:
what about the flush?

Reply:
I added it